### PR TITLE
VST3 factory presets

### PR DESCRIPTION
**Pull request by vsverchinsky**
Resolves: #3860

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/effects/VST3/VST3Wrapper.cpp [hunk 1]
```diff
@@ -240,12 +240,9 @@ class ComponentHandler : public Steinberg::Vst::IComponentHandler
...
          mWrapper.mEditController->getParameterInfo(i, info);
 
-         if (info.flags & (ParameterInfo::kIsHidden | ParameterInfo::kIsProgramChange))
+         if ((info.flags & ParameterInfo::kIsReadOnly) != 0)
             continue;
-
-         if ((info.flags & ParameterInfo::kCanAutomate)  == 0)
-            continue;
-
+
```
Comment (Paul-Licameli) [inline src/effects/VST3/VST3Wrapper.cpp:275]: It's not obvious locally in the code here, but I understand a consequence of this is that changes of factory preset will "dirty" the project so you are prompted to save it when you close.  Am I right?  Is there another reason for this?

File: src/effects/VST3/VST3Instance.cpp [hunk 2]
```diff
@@ -153,7 +154,7 @@ bool VST3Instance::ProcessFinalize() noexcept
 
 bool VST3Instance::ProcessInitialize(EffectSettings &settings, double sampleRate, ChannelNames chanMap)
 {
-   if(mWrapper->Initialize(settings, sampleRate, Steinberg::Vst::kRealtime, mProcessingBlockSize))
+   if(mWrapper->Initialize(settings, sampleRate, Steinberg::Vst::kOffline, mProcessingBlockSize))
```
Comment (Paul-Licameli) [inline src/effects/VST3/VST3Instance.cpp:157]: I understand the second of the six commits, except for this unexplained change.  What does it accomplish?

File: src/effects/VST3/VST3Wrapper.h [hunk 3]
```diff
@@ -114,6 +123,10 @@ class VST3Wrapper
    void BeginParameterEdit(EffectSettingsAccess& access);
    void EndParameterEdit();
 
+   //! Returns an array of factory preset ids.
+   //! Safe to call before `InitilizeComponents`
```
Comment (Paul-Licameli) [inline src/effects/VST3/VST3Wrapper.h:127]: "InitializeComponents"

File: src/effects/VST3/VST3Wrapper.cpp [hunk 4]
```diff
@@ -544,28 +552,61 @@ void VST3Wrapper::StoreSettings(EffectSettings& settings) const
...
+   using namespace Steinberg;
+
+   auto fileStream = owned(Vst::FileStream::open(presetId.c_str(), "rb"));
+   if(!fileStream)
+      throw FileException(FileException::Cause::Open, presetId);
+
+   if(!LoadPresetFromStream(fileStream))
+   {
+      throw SimpleMessageBoxException(
+         ExceptionType::Internal,
```
Comment (Paul-Licameli) [inline src/effects/VST3/VST3Wrapper.cpp:690]: Is `Internal` the right type for this and the other new exception, not `BadEnvironment`?  If this failure occurs, do we want Sentry data about it or not?

`FileException` does use `Internal` but see the comment in its constructor explaining that as an unusual decision.

File: src/effects/VST3/VST3Utils.cpp [hunk 5]
```diff
@@ -122,6 +122,27 @@ bool VST3Utils::ParseAutomationParameterKey(const wxString& key, Steinberg::Vst:
    return false;
 }
 
+wxString VST3Utils::MakeFactoryPresetID(Steinberg::Vst::UnitID unitId, Steinberg::int32 programIndex)
+{
+   return wxString::Format("%d:%d",
```
Comment (Paul-Licameli) [inline src/effects/VST3/VST3Utils.cpp:127]: Are you certain ':' will not occur in either of the substituted strings?  The inverse function assumes this.

If you doubt it, maybe the utility functions `wxSplit` and `wxJoin` can be used, to easily escape  certain characters.

File: src/effects/VST3/VST3Wrapper.cpp [hunk 6]
```diff
@@ -562,8 +563,14 @@ void VST3Wrapper::InitializeComponents()
       mEditController->setComponentState(&stateStream);
    }
 
-   mDefaultSettings = MakeSettings();
-   StoreSettings(mDefaultSettings);
+   {
+      auto cache = GetCache(mEffectClassInfo.ID());
+      if(!cache->defaultSettings.has_value())
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/effects/VST3/VST3Wrapper.cpp:568]: In fact `GetCache` is not now guaranteeing a non-null return, so please null check this.

**Review by Paul-Licameli — DISMISSED** [review]

File: src/effects/VST3/VST3Wrapper.cpp [hunk 7]
```diff
@@ -28,6 +28,32 @@ constexpr auto processorStateKey  = wxT("ProcessorState");
...
+      return &it->second;
+   return nullptr;
+}
+
+VST3PluginCache& CreateCache(const VST3::UID& effectUid)
+{
+   const auto key = effectUid.toString();
+   assert(sVST3PluginCache.find(key) == sVST3PluginCache.end());
+   const auto result = sVST3PluginCache.insert(std::pair { key, VST3PluginCache {} });
+   return result.first->second;
```
Comment (Paul-Licameli) [inline src/effects/VST3/VST3Wrapper.cpp:55]: Why not simply combine the two functions ?  Just let `GetCache` return a reference to `sVST3PluginCache[key]` and that will create and default-initialize a member that isn't there already.

Then the assertion isn't needed, and you know I don't like assertions without comments:  when I review an assertion,. I'm not satisfied until I also review all uses of the function.  Less burden of review without it.

File: src/effects/VST3/VST3Wrapper.cpp [hunk 8]
```diff
@@ -416,6 +442,66 @@ VST3Wrapper::VST3Wrapper(VST3::Hosting::Module& module, const VST3::Hosting::Cla
    mEditController = editController;
 
    mEditController->initialize(&AudacityVst3HostApplication::Get());
+
+   if(GetCache(mEffectClassInfo.ID()) == nullptr)
```
Comment (Paul-Licameli) [inline src/effects/VST3/VST3Wrapper.cpp:447]: To make my last suggestion work, defaulted `VST3PluginCache` must be distinguishable from one that was initialized as in the following code, so you can write another test here.

File: src/effects/VST3/VST3Wrapper.cpp [hunk 9]
```diff
@@ -416,6 +442,66 @@ VST3Wrapper::VST3Wrapper(VST3::Hosting::Module& module, const VST3::Hosting::Cla
...
+         Vst::ParameterInfo parameterInfo {};
+         if(mEditController->getParameterInfo(parameterIndex, parameterInfo) != kResultOk)
+            continue;
+
+         if(parameterInfo.unitId != Vst::kRootUnitId)
+            continue;
+         if((parameterInfo.flags & Vst::ParameterInfo::kIsProgramChange) == 0)
+            continue;
+
+         cache.rootUnitProgramChangeParameterID = parameterInfo.id;
```
Comment (Paul-Licameli) [inline src/effects/VST3/VST3Wrapper.cpp:466]: ... and I think this assignment makes that distinguishability.

File: src/effects/VST3/VST3Wrapper.cpp [hunk 10]
```diff
@@ -564,7 +681,7 @@ void VST3Wrapper::LoadPreset(const wxString& presetId)
    {
       throw SimpleMessageBoxException(
          ExceptionType::Internal,
-         XO("Unable to apply VST3 preset file %s").Format(presetId),
+         XO("Unable to apply VST3 preset %s").Format(presetId),
```
Comment (Paul-Licameli) [inline src/effects/VST3/VST3Wrapper.cpp:691]: This is a change of a user facing string that I suppose you must tell design about, if there is any important opinion about string consistency.

**Comment by Paul-Licameli** [comment]
I'm done reviewing.  Mostly good.  But I had some questions.

File: src/effects/VST3/VST3Wrapper.cpp (hunk 1 above)
Comment (vsverchinsky) [reply src/effects/VST3/VST3Wrapper.cpp:275]: It is correct, though formally there could be other types of parameter that aren't automatable or hidden. For such parameters it is correct to "dirty" the project too I think.

File: src/effects/VST3/VST3Instance.cpp (hunk 2 above)
Comment (vsverchinsky) [reply src/effects/VST3/VST3Instance.cpp:157]: Some VST3 plugins can offer better processig quality in `kOffline` mode. I think it used to be `kOffline` some time before refactoring, but I've made a typo

File: src/effects/VST3/VST3Wrapper.cpp (hunk 4 above)
Comment (vsverchinsky) [reply src/effects/VST3/VST3Wrapper.cpp:690]: > Is Internal the right type for this and the other new exception, not BadEnvironment? If this failure occurs, do we want Sentry data about it or not?

Not sure how to answer the first question, but I don't think that it's a good idea to send preset loading failures to sentry. `LoadPresetFromStream(fileStream)` load preset file from the disc, and user may attempt to load preset that was genrated by older version of same plugin, or even by another plugin.

File: src/effects/VST3/VST3Wrapper.cpp (hunk 7 above)
Comment (vsverchinsky) [reply src/effects/VST3/VST3Wrapper.cpp:55]: The answer is in VST3Wrapper constructor - cache initialization process. If `GetCache` and `CreateCahce` were combined I had to add an initialization state. Reference type might have convinced you that return value isn't nullptr (though that can be violated too), and the code is formally correct, but you must make sure that the state of returned object is correct too. I don't think that it worth changing.

File: src/effects/VST3/VST3Wrapper.cpp (hunk 8 above)
Comment (vsverchinsky) [reply src/effects/VST3/VST3Wrapper.cpp:447]: Yes, and all other places should check that too. Not obvious what should happen if `GetCache` returns not initialized object

File: src/effects/VST3/VST3Wrapper.cpp (hunk 10 above)
Comment (vsverchinsky) [reply src/effects/VST3/VST3Wrapper.cpp:691]: Right, thanks!

File: src/effects/VST3/VST3Wrapper.cpp (hunk 9 above)
Comment (vsverchinsky) [reply src/effects/VST3/VST3Wrapper.cpp:466]: Yes, the meaning for `kNoProgramListId` is that plugin does not have program list parameter, which is fine.

File: src/effects/VST3/VST3Instance.cpp (hunk 2 above)
Comment (Paul-Licameli) [reply src/effects/VST3/VST3Instance.cpp:157]: So then, will you revert that change or comment that you are keeping it?

File: src/effects/VST3/VST3Wrapper.cpp (hunk 4 above)
Comment (Paul-Licameli) [reply src/effects/VST3/VST3Wrapper.cpp:690]: If there should not be Sentry reporting for this error, then Internal is the wrong choice.  See this other example:

```
         GuardedCall(
            [&message, rc] {
            throw SimpleMessageBoxException{ rc != SQLITE_FULL ? ExceptionType::Internal : ExceptionType::BadEnvironment,
               message, XO("Warning"), "Error:_Disk_full_or_not_writable" }; },
            SimpleGuard<void>{},
            [this](AudacityException * e) {
               // This executes in the main thread.
               if (mCallback)
                  mCallback();
               if (e)
                  e->DelayedHandlerAction();
            }
         );
```

If Sqlite fails but tells the program it's exhaustion of device space, then we don't need to collect a report.  Only if Sqlite fails for some other various reason do we want a report.

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]
Still I must request one change, to pass the correct exception type and avoid unnecessary error reports.

Besides that, just one spelling mistake in a comment, and resolve the question whether `kOffline` is right.

File: src/effects/VST3/VST3Instance.cpp (hunk 2 above)
Comment (vsverchinsky) [reply src/effects/VST3/VST3Instance.cpp:152]: No, I think `kOffline` fits better for destructive processing. `RealtimeInitialize` uses `kRealtime` mode.

File: src/effects/VST3/VST3Wrapper.cpp (hunk 10 above)
Comment (vsverchinsky) [reply src/effects/VST3/VST3Wrapper.cpp]: Reverted that line

**Comment by DavidBailes** [comment]
@vsverchinsky . It's great to see that this is being worked on so that keyboard users can access the factory presets of vst3 plugins:
https://github.com/audacity/audacity/issues/3414

I've tested the plugin mentioned by that issue: https://www.auburnsounds.com/products/Couture.html, and can confirm that the factory presets are now available.
However, I've tested with TDR Kotelnikov  (https://www.tokyodawn.net/tdr-kotelnikov/), and TDR VOS SlickEQ (https://www.tokyodawn.net/tdr-vos-slickeq/), and the presets are still not available for these. (They are using Reaper.)

**Review by Paul-Licameli — APPROVED** [review]

**Comment by Paul-Licameli** [comment]
All my questions are resolve now, BUT, the latest run failed to build for Ubuntu.

But those looked like spurious failures caused by intermittent network problems, so just started a rerun of the failed jobs.  If that passes, then this can be merge.

**Comment by p0nce** [comment]
You can use Graillon free VST3 for VST3 program lists, we are actually getting support issues from Audacity users :) https://www.auburnsounds.com/products/Graillon.html

**Comment by vsverchinsky** [comment]
> @vsverchinsky . It's great to see that this is being worked on so that keyboard users can access the factory presets of vst3 plugins: #3414
> 
> I've tested the plugin mentioned by that issue: https://www.auburnsounds.com/products/Couture.html, and can confirm that the factory presets are now available. However, I've also tested with TDR Kotelnikov (https://www.tokyodawn.net/tdr-kotelnikov/), and TDR VOS SlickEQ (https://www.tokyodawn.net/tdr-vos-slickeq/), and the presets are still not available for these. (They are using Reaper.)

@DavidBailes I've updated PR and plugins mentioned should also list factory programs in the "settings & presets" menu. Though one issue remains - there could be more than one menu entry named 'Defaults', and moreover they could be same, but I'm not sure what to do about it right now (the problem is that for some plugins that may be not true, I think)

**Comment by DavidBailes** [comment]
> @DavidBailes I've updated PR and plugins mentioned should also list factory programs in the "settings & presets" menu. Though one issue remains - there could be more than one menu entry named 'Defaults', and moreover they could be same, but I'm not sure what to do about it right now (the problem is that for some plugins that may be not true, I think)

@vsverchinsky . Thanks, I can confirm that the problem with those two plugins is fixed.
The issue with entries for defaults and default is a minor issue (I notice that it also occurs for vst2 plugins).

**Review by Paul-Licameli — APPROVED** [review]
I have reviewed what is new since last review.  I have no comments.

**Pull request merged**